src/app_engine/join_page.py

```diff
@@ -200,3 +200,90 @@ def post(self, room_id):
         result['is_initiator'])
     logging.info('User ' + client_id + ' joined room ' + room_id +
                  ' with state ' + result['room_state'])
+
+
+class AckPage(webapp2.RequestHandler):
+  """Handles ack requests."""
+
+  def write_response(self, result):
+    self.response.write(json.dumps({
+        'result': result,
+    }))
+
+  def report_error(self, error):
+    self.write_response(error)
+
+  def handle_ringing(self, msg):
+    """Handles a callee acknowledging the incoming call."""
+    if (not util.has_msg_field(msg, constants.PARAM_CALLEE_GCM_ID, basestring)
+        or not util.has_msg_field(msg, constants.PARAM_ROOM_ID, basestring)):
+      self.report_error(constants.RESPONSE_INVALID_ARGUMENT)
+      return
+
+    callee_gcm_id = msg[constants.PARAM_CALLEE_GCM_ID]
+    room_id = msg[constants.PARAM_ROOM_ID]
+
+    call_room = room.get_room(self.request.host_url, room_id)
+    if not call_room:
+      logging.warning('Room ' + room_id + ' does not exist when trying to ' +
+                      'acknowledge a call. Callee gcm id: ' + callee_gcm_id)
+      self.report_error(constants.RESPONSE_INVALID_ROOM)
+      return
+
+    # Look up and validate callee by gcm id.
+    # TODO(chuckhays): Once registration is enabled, turn on restriction
+    # to only return verified records.
+    callee_records = gcmrecord.GCMRecord.get_associated_records_for_gcm_id(
+        callee_gcm_id, False)
+    if len(callee_records) < 1:
+      self.report_error(constants.RESPONSE_INVALID_CALLEE)
+      return
+    callee_id = callee_records[0].user_id
+
+    # Notify caller that callee has acknowledged the call and is ringing.
+    # There should only be 1 client in the room, the caller.
+    room_state = call_room.get_room_state()
+    if room_state != room.Room.STATE_WAITING:
+      logging.warning('Room ' + room_id + ' is not in WAITING state ' +
```

Comment:
That probably makes sense, the request is only useful when in the waiting state.